crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+                        .context("type mismatch for option")?;
+
+                    Ok::<_, Error>(value)
+                })
+                .transpose()?;
+
+            Ok(Val::Option(Option {
+                ty: handle.clone(),
+                discriminant: if value.is_none() { 0 } else { 1 },
+                value: Box::new(value.unwrap_or(Val::Unit)),
+            }))
+        } else {
+            Err(anyhow!("cannot make option from {} type", self.desc()))
+        }
+    }
+
+    /// Instantiate this type (which must be a `Type::Expected`) with the specified `value`.
+    pub fn new_expected(&self, value: Result<Val, Val>) -> Result<Val> {
+        if let Type::Expected(handle) = self {
+            Ok(Val::Expected(Expected {
+                ty: handle.clone(),
+                discriminant: if value.is_ok() { 0 } else { 1 },
+                value: Box::new(match value {
+                    Ok(value) => {
+                        handle
+                            .ok()
+                            .check(&value)
+                            .context("type mismatch for ok case of expected")?;
+                        value
+                    }
+                    Err(value) => {
+                        handle
+                            .err()
+                            .check(&value)
+                            .context("type mismatch for err case of expected")?;
+                        value
+                    }
+                }),
+            }))
+        } else {
+            Err(anyhow!("cannot make expected from {} type", self.desc()))
+        }
+    }
+
+    /// Instantiate this type (which must be a `Type::Flags`) with the specified flag `names`.
+    pub fn new_flags(&self, names: &[&str]) -> Result<Val> {
+        if let Type::Flags(handle) = self {
+            let map = handle
+                .names()
+                .enumerate()
+                .map(|(index, name)| (name, index))
+                .collect::<HashMap<_, _>>();
+
+            let mut values = vec![0_u32; values::u32_count_for_flag_count(handle.names().len())];
+
+            for name in names {
+                let index = map
+                    .get(name)
+                    .ok_or_else(|| anyhow!("unknown flag: {name}"))?;
+                values[index / 32] |= 1 << (index % 32);
+            }
+
+            Ok(Val::Flags(Flags {
+                ty: handle.clone(),
+                count: u32::try_from(map.len())?,
+                value: values.into(),
+            }))
+        } else {
+            Err(anyhow!("cannot make flags from {} type", self.desc()))
+        }
+    }
+
+    /// Retrieve any types nested within this type.
+    ///
+    /// For example, this will return the field types in order of declaration for a record type.  It will return
+    /// the element type for a list type and nothing for non-composite types like bool, u8, and string.
+    pub fn nested(&self) -> Box<[Type]> {
```

Comment:
Is this still necessary? This seems like a bit of an odd query that I'm not sure what it would be used for

Reply:
Please see my comment above about ergonomics and how it is used in the dynamic.rs tests.

Reply:
@alexcrichton any further concerns about this method?  I'm going to mark this "resolved" for now, but feel free to unresolve it.